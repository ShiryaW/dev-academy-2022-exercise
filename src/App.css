.App {
  text-align: center;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.data-grid {
  width: 96vw;
  margin: 0 auto;
  height: 70vh;
}

#toolbar {
  min-width: 929px;
  margin-bottom: 2vh;
}

#chart {
  margin: 0 auto;
  display: table;
}