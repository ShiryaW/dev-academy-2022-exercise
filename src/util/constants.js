export const SERVICE_URL = "http://localhost";
export const SERVICE_PORT = 8080;
export const ALL = "all";

export const ERRORS = {
  MISSING_DATE: "Date missing",
  MALFORMED_INPUT: "Malformed input",
};

export const FARM_NAMES = {
  FRIMAN: "Friman Metsola Collective",
  PARTIALTECH: "PartialTech Research Farm",
  NOORA: "Noora's Farm",
  OSSI: "Organic Ossi's Impact That Lasts Plantation",
};

export const FARM_IDS = {
  [FARM_NAMES.FRIMAN]: 1,
  [FARM_NAMES.PARTIALTECH]: 2,
  [FARM_NAMES.NOORA]: 3,
  [FARM_NAMES.OSSI]: 4,
};

export const FIELD_KEYS = {
  DATETIME: "datetime",
  DATETIME_RAW: "datetime_raw",
  FARM_ID: "farm_id",
  ID: "id",
  LOCATION: "location",
  SENSOR_TYPE: "sensor_type",
  VALUE: "value",
};

export const SENSOR_TYPES = {
  TEMPERATURE: "temperature",
  RAINFALL: "rainfall",
  PH: "pH",
};
