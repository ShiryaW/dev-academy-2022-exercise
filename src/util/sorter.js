export const datesort = (date1, date2) => {
  const [day1, month1, year1] = parseDayMonthYear(date1);
  const [day2, month2, year2] = parseDayMonthYear(date2);

  if (year1 < year2) return -1;
  else if (year1 === year2) {
    if (month1 < month2) return -1;
    else if (month1 === month2) {
      if (day1 < day2) return -1;
      else if (day1 === day2) return 0;
      else return 1;
    } else return 1;
  } else return 1;
};

export const parseDayMonthYear = (date) => {
  return date
    .split(",")[0]
    .split(".")
    .map((n) => parseInt(n));
};
