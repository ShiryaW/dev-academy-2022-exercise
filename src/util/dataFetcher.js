import { FARM_IDS } from "./constants";

export const fetchFarmNames = async () => {
  const body = await fetch(`v1/farms`).then((res) => res.json());
  return body.map(({ farm_id, name }) => {
    return {
      id: farm_id,
      name,
    };
  });
};

export const getFarmStats = async (farm) => {
  return await fetch(`/v1/farms/${farm}/stats`).then((res) => res.json());
};

export const getStatsForAll = async () => {
  const promises = Object.values(FARM_IDS).map(
    async (id) =>
      await getFarmStats(id).then(({ measurements }) => measurements)
  );
  const measurements = await Promise.all(promises);
  return { measurements: measurements.flat() };
};
