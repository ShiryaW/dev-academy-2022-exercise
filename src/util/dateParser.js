import { ERRORS } from "./constants";

export const parseDate = (iso) => {
  // Input format: 2019-01-01T23:10:20.551Z
  if (!iso) throw new Error(ERRORS.MISSING_DATE);
  let date, time, dateArr, timeArr;

  try {
    [date, time] = iso.slice(0, -1).split("T");
    dateArr = date.split("-");
    timeArr = time.split(":");
  } catch (e) {
    throw new Error(ERRORS.MALFORMED_INPUT);
  }

  if (!date || !time || dateArr.length !== 3 || timeArr.length !== 3)
    throw new Error(ERRORS.MALFORMED_INPUT);

  return `${dateArr[2]}.${dateArr[1]}.${dateArr[0]}, ${timeArr[0]}:${
    timeArr[1]
  }:${Math.round(timeArr[2]).toString().padStart(2, "0")}`;
};

export const getDate = (iso) => {
  let result = ""; // fallback on an empty string if validation fails

  try {
    result = parseDate(iso);
  } catch (e) {
    if (e.message === ERRORS.MISSING_DATE) return e.message; // display placeholder text if there is no date
  }

  return result;
};
