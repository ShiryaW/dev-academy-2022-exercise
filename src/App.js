import "./App.css";
import PropTypes from "prop-types";
import React, { useEffect, useState } from "react";
import { DataGrid } from "@mui/x-data-grid";
import { Toolbar } from "./Toolbar.js";
import {
  fetchFarmNames,
  getFarmStats,
  getStatsForAll,
} from "./util/dataFetcher";
import { ALL, FARM_IDS, FIELD_KEYS, SENSOR_TYPES } from "./util/constants";
import { getDate } from "./util/dateParser";
import { Chart } from "./Chart";
import { datesort } from "./util/sorter";

/* NOTES TO SELF:
  What kind of questions can we answer?
  1) Is there a correlation between rainfall/pH/temperature on a given farm?
  2) Which farm gets the most rainfall/has the highest pH/highest temperature?
  3) How has the value of X progressed over time on a given farm?
  4) What are the average values of X on each farm?
 */

export const App = () => {
  const maxColWidth = 400;
  const [farmNames, setFarmNames] = useState([]);
  const [selectedFarm, setSelectedFarm] = useState(ALL);
  const [chartTable, setChartTable] = useState({ measurements: [] });
  const [rows, setRowData] = useState([]);

  useEffect(async () => {
    const names = await fetchFarmNames();
    if (names.error) return;
    setFarmNames(names);
  }, []);

  const columns = [
    {
      field: FIELD_KEYS.LOCATION,
      headerName: "Location",
      minWidth: 200,
      flex: maxColWidth,
    },
    {
      field: FIELD_KEYS.DATETIME,
      headerName: "Time of measurement",
      minWidth: 200,
      sortComparator: datesort,
      flex: maxColWidth,
    },
    {
      field: FIELD_KEYS.SENSOR_TYPE,
      headerName: "Type",
      minWidth: 150,
      flex: maxColWidth,
      description:
        "The type of measurement taken. Possible values are rainfall, pH or temperature.",
      sortable: false,
    },
    {
      field: FIELD_KEYS.VALUE,
      headerName: "Value",
      minWidth: 150,
      flex: maxColWidth,
    },
  ];

  const handleButtonClick = async ({ target }) => {
    const targetFarm = FARM_IDS[target.textContent];

    if (targetFarm) {
      setSelectedFarm(target.textContent);
      await getFarmStats(targetFarm).then((data) => generateRows(data));
    } else {
      setSelectedFarm(ALL);
      await getStatsForAll().then((data) => generateRows(data));
    }
  };

  const generateRows = async ({ measurements }) => {
    const rowData = measurements.map((row, idx) => ({
      ...row,
      id: idx,
      datetime: getDate(row.datetime),
      datetime_raw: row.datetime,
    }));
    setChartTable({ measurements: rowData });
    setRowData([...rowData]);
  };

  return (
    <>
      <Toolbar farms={farmNames} onButtonClick={handleButtonClick} />
      {rows.length > 0 ? (
        <Chart
          data={chartTable}
          sensorType={SENSOR_TYPES.TEMPERATURE}
          selectedFarm={selectedFarm}
        />
      ) : (
        <Chart data={chartTable} />
      )}
      <div className="data-grid">
        <DataGrid columns={columns} rows={rows} />
      </div>
    </>
  );
};

App.propTypes = {
  serviceUrl: PropTypes.string.isRequired,
  port: PropTypes.number.isRequired,
};
