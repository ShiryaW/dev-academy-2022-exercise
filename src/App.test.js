import React from "react";
import { shallow } from "enzyme";
import { App } from "./App";
import { SERVICE_PORT, SERVICE_URL } from "./util/constants";

const createApp = (props) =>
  shallow(
    <App {...{ ...props, serviceUrl: SERVICE_URL, port: SERVICE_PORT }} />
  );

describe("App", () => {
  it("renders the grid component", () => {
    const app = createApp({});

    expect(app).toMatchSnapshot();
  });
});
