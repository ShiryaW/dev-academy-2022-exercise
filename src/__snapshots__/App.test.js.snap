// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the grid component 1`] = `
<Fragment>
  <Toolbar
    farms={Array []}
    onButtonClick={[Function]}
  />
  <Chart
    data={
      Object {
        "measurements": Array [],
      }
    }
  />
  <div
    className="data-grid"
  >
    <Memo(ForwardRef)
      columns={
        Array [
          Object {
            "field": "location",
            "flex": 400,
            "headerName": "Location",
            "minWidth": 200,
          },
          Object {
            "field": "datetime",
            "flex": 400,
            "headerName": "Time of measurement",
            "minWidth": 200,
          },
          Object {
            "description": "The type of measurement taken. Possible values are rainfall, pH or temperature.",
            "field": "sensor_type",
            "flex": 400,
            "headerName": "Type",
            "minWidth": 150,
            "sortable": false,
          },
          Object {
            "field": "value",
            "flex": 400,
            "headerName": "Value",
            "minWidth": 150,
          },
        ]
      }
      rows={Array []}
    />
  </div>
</Fragment>
`;
